plugins/soundsourcemediafoundation/audiosourcemediafoundation.cpp

```diff
@@ -0,0 +1,558 @@
+        return false;
+    }
+
+    // MSDN for this attribute says that if bps is 8, samples are unsigned.
+    // Otherwise, they're signed (so they're signed for us as 16 bps). Why
+    // chose to hide this rather useful tidbit here is beyond me -bkgood
+    hr = m_pAudioType->SetUINT32(MF_MT_AUDIO_BITS_PER_SAMPLE, sizeof(m_leftoverBuffer[0]) * 8);
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to set bits per sample";
+        return false;
+    }
+
+    hr = m_pAudioType->SetUINT32(MF_MT_AUDIO_BLOCK_ALIGNMENT,
+            frames2samples(sizeof(m_leftoverBuffer[0])));
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to set block alignment";
+        return false;
+    }
+
+    hr = m_pAudioType->SetUINT32(MF_MT_AUDIO_NUM_CHANNELS, kNumChannels);
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to set number of channels";
+        return false;
+    }
+
+    hr = m_pAudioType->SetUINT32(MF_MT_AUDIO_SAMPLES_PER_SECOND, kSampleRate);
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to set sample rate";
+        return false;
+    }
+
+    // Set this type on the source reader. The source reader will
+    // load the necessary decoder.
+    hr = m_pReader->SetCurrentMediaType(MF_SOURCE_READER_FIRST_AUDIO_STREAM,
+    NULL, m_pAudioType);
+
+    // the reader has the media type now, free our reference so we can use our
+    // pointer for other purposes. Do this before checking for failure so we
+    // don't dangle.
+    safeRelease(&m_pAudioType);
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to set media type";
+        return false;
+    }
+
+    // Get the complete uncompressed format.
+    hr = m_pReader->GetCurrentMediaType(MF_SOURCE_READER_FIRST_AUDIO_STREAM,
+            &m_pAudioType);
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to retrieve completed media type";
+        return false;
+    }
+
+    // Ensure the stream is selected.
+    hr = m_pReader->SetStreamSelection(MF_SOURCE_READER_FIRST_AUDIO_STREAM,
+            true);
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to select first audio stream (again)";
+        return false;
+    }
+
+    UINT32 leftoverBufferSize = 0;
+    hr = m_pAudioType->GetUINT32(MF_MT_SAMPLE_SIZE, &leftoverBufferSize);
+    if (FAILED(hr)) {
+        qWarning() << "SSMF: failed to get buffer size";
+        return false;
+    }
+    m_leftoverBufferSize = leftoverBufferSize;
+    m_leftoverBufferSize /= sizeof(sample_type); // convert size in bytes to sizeof(sample_type)
+    m_leftoverBuffer = new sample_type[m_leftoverBufferSize];
+
+    return true;
+}
+
+/**
+ * Copies min(destFrames, srcFrames) frames to dest from src. Anything leftover
+ * is moved to the beginning of m_leftoverBuffer, so empty it first (possibly
+ * with this method). If src and dest overlap, I'll hurt you.
+ */
+void AudioSourceMediaFoundation::copyFrames(sample_type *dest, size_t *destFrames,
```

Comment:
We can probably improve this by using https://github.com/mixxxdj/mixxx/blob/master/src/util/pa_ringbuffer.c
I do not want to put this work on you. So please left a TODO(XXX) here.